compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -510,7 +510,7 @@ impl server::FreeFunctions for Rustc<'_, '_> {
 
     fn emit_diagnostic(&mut self, diagnostic: Diagnostic<Self::Span>) {
         let message = rustc_errors::DiagnosticMessage::from(diagnostic.message);
-        let mut diag: DiagnosticBuilder<'_, rustc_errors::ErrorGuaranteed> =
+        let mut diag: DiagnosticBuilder<'_, ()> =
```

Comment:
I see why this fixed the issue, but why are the other changes necessary?

Reply:
You mean in the second commit? While I was looking at this I did an audit of all the `DiagnosticBuilder::new` calls to make sure there weren't any other problems. That's when I noticed the non-generic `IntoDiagnostic` impls, and since they are kind of related I decided to do them in the same PR (albeit in a different commit).

Reply:
Some of them are explicitly setting error codes. It doesn't really make sense to allow them to be anything but an error. I guess that's a limitation of the `IntoDiagnostic` API

I don't think they should be made generic if they are always meant to be errors

Reply:
#119097 had a lot of stuff about this, see [this comment](https://github.com/rust-lang/rust/pull/119097#issuecomment-1865181667) for a summary. In short, all derived `IntoDiagnostic` impls have to be generic. For consistency, we decided that all hand-written ones should be too. That way, for all diagnostics, its the emission site that dictates the level. But four of them were overlooked.

Reply:
Right, we should probably create a rustc lint for t

Reply:
Maybe the `IntoDiagnostic` doc comment should explain that impls should always be generic, and why?

Reply:
I've added that to my todo list, I'll get to it next week.